libcnb/src/tracing.rs

```diff
@@ -0,0 +1,136 @@
+use libcnb_data::buildpack::Buildpack;
+use opentelemetry::{
+    global,
+    trace::{Span as SpanTrait, Status, Tracer, TracerProvider as TracerProviderTrait},
+    KeyValue,
+};
+use opentelemetry_sdk::{
+    trace::{Config, Span, TracerProvider},
+    Resource,
+};
+use std::path::Path;
+
+const TELEMETRY_EXPORT_ROOT: &str = "/tmp/cnb-telemetry";
+
+pub(crate) struct BuildpackTrace {
+    provider: TracerProvider,
+    span: Span,
+}
+
+pub(crate) fn start_trace(buildpack: &Buildpack, phase_name: &'static str) -> BuildpackTrace {
+    let trace_name = format!("{}-{phase_name}", buildpack.id.replace(['/', '.'], "_"));
+    let tracing_file_path = Path::new(TELEMETRY_EXPORT_ROOT).join(format!("{trace_name}.jsonl"));
+
+    // Ensure tracing file path parent exists by creating it.
+    if let Some(parent_dir) = tracing_file_path.parent() {
+        let _ = std::fs::create_dir_all(parent_dir);
+    }
+    let exporter = match std::fs::File::options()
+        .create(true)
+        .append(true)
+        .open(&tracing_file_path)
+    {
+        Ok(file) => opentelemetry_stdout::SpanExporter::builder()
+            .with_writer(file)
+            .build(),
+        Err(_) => opentelemetry_stdout::SpanExporter::builder()
+            .with_writer(std::io::sink())
+            .build(),
+    };
+
+    let lib_name = option_env!("CARGO_PKG_NAME").unwrap_or("libcnb");
+    let lib_version = option_env!("CARGO_PKG_VERSION");
+
+    let provider = TracerProvider::builder()
+        .with_simple_exporter(exporter)
+        .with_config(Config::default().with_resource(Resource::new(vec![
+            KeyValue::new("service.name", lib_name),
+            KeyValue::new("service.version", lib_version.unwrap_or_default()),
+        ])))
+        .build();
+
+    global::set_tracer_provider(provider.clone());
+
+    let tracer = provider.versioned_tracer(lib_name, lib_version, None as Option<&str>, None);
+
+    let mut span = tracer.start(trace_name);
+    span.set_attributes(vec![
+        KeyValue::new("buildpack_id", buildpack.id.to_string().clone()),
+        KeyValue::new("buildpack_name", buildpack.name.clone().unwrap_or_default()),
+        KeyValue::new("buildpack_version", buildpack.version.to_string()),
+        KeyValue::new(
+            "buildpack_homepage",
+            buildpack.homepage.clone().unwrap_or_default(),
+        ),
+    ]);
+    BuildpackTrace { provider, span }
+}
+
+impl BuildpackTrace {
+    pub(crate) fn set_error(&mut self, err: &dyn std::error::Error) {
+        self.span.set_status(Status::error(err.to_string()));
+        self.span.record_error(err);
+    }
+    pub(crate) fn add_event(&mut self, name: &'static str) {
+        self.span.add_event(name, vec![]);
+    }
+}
+
+impl Drop for BuildpackTrace {
+    fn drop(&mut self) {
+        self.span.end();
+        self.provider.force_flush();
+        global::shutdown_tracer_provider();
+    }
+}
+
+#[cfg(test)]
+mod tests {
```

Comment:
I think we need some libcnb-test integration tests for this - perhaps by having the integration test run two buildpacks, with the second buildpack printing the output file contents from `/tmp`? (Since I presume there is no other easy way to get the contents out locally)

Reply:
Added in ab24d40370f85bf88735399e14b24a00c8fbc40b.